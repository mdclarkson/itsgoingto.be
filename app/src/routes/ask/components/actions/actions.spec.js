import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import WithRouter from '../../../../../test-utils/with-router'
import Actions from './actions'

const mockHistory = {
  push: jest.fn()
}

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => mockHistory,
}));

const props = {
  hasQuestion: true,
  canSubmitPoll: true,
  postPoll: jest.fn(() => Promise.resolve({ identifier: 'jdH93HS' }))
}

describe('(Route) Ask', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('(Component) Actions', () => {
    describe('(Action) Create Poll click', () => {
      it('should call postPoll', () => {
        render(<Actions {...props} />, {
          wrapper: WithRouter
        })

        const button = screen.getByTestId('button-Create-Poll')
        fireEvent.click(button)
        expect(props.postPoll).toBeCalled()
      })

      it('should redirect with poll identifier on successful postPoll and return false', async () => {
        render(<Actions {...props} />, {
          wrapper: WithRouter
        })

        const button = screen.getByTestId('button-Create-Poll')
        await fireEvent.click(button)
        expect(mockHistory.push).toBeCalledWith('/jdH93HS')
      })

      it('should not redirect failed postPoll and return true', async () => {
        render(<Actions {...props} postPoll={jest.fn(() => Promise.resolve(false))} />, {
          wrapper: WithRouter
        })

        const button = screen.getByTestId('button-Create-Poll')
        await fireEvent.click(button)
        expect(mockHistory.push).not.toBeCalled()
      })
    })

    describe('(Render)', () => {
      describe('when hasQuestion is true', () => {
        it('matches snapshot', () => {
          const { asFragment } = render(<Actions {...props} hasQuestion />)
          expect(asFragment()).toMatchSnapshot()
        })
      })

      describe('when hasQuestion is false', () => {
        it('matches snapshot', () => {
          const { asFragment } = render(<Actions {...props} hasQuestion={false} />)
          expect(asFragment()).toMatchSnapshot()
        })
      })

      describe('when canSubmitPoll is true', () => {
        it('matches snapshot', () => {
          const { asFragment } = render(<Actions {...props} canSubmitPoll />)
          expect(asFragment()).toMatchSnapshot()
        })
      })

      describe('when canSubmitPoll is false', () => {
        it('matches snapshot', () => {
          const { asFragment } = render(<Actions {...props} canSubmitPoll={false} />)
          expect(asFragment()).toMatchSnapshot()
        })
      })
    })
  })
})
