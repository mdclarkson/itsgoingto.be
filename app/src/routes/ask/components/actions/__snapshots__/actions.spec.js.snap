// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Route) Ask (Component) Actions (Render) when canSubmitPoll is false matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="actions hideable"
    >
      <button
        class="btn pull-right disabled"
        data-testid="button-Create-Poll"
        disabled=""
        type="button"
      >
        <span>
          Create Poll
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`(Route) Ask (Component) Actions (Render) when canSubmitPoll is true matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="actions hideable"
    >
      <button
        class="btn pull-right"
        data-testid="button-Create-Poll"
        type="button"
      >
        <span>
          Create Poll
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`(Route) Ask (Component) Actions (Render) when hasQuestion is false matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="actions hideable gone"
    >
      <button
        class="btn pull-right"
        data-testid="button-Create-Poll"
        type="button"
      >
        <span>
          Create Poll
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`(Route) Ask (Component) Actions (Render) when hasQuestion is true matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="actions hideable"
    >
      <button
        class="btn pull-right"
        data-testid="button-Create-Poll"
        type="button"
      >
        <span>
          Create Poll
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
